clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3057,10 +3092,13 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
   Expr *ZDimExpr = AL.isArgExpr(2) ? AL.getArgAsExpr(2)
                                    : SetDefaultValue(S, AL, AL.getLoc());
 
+  // __attribute__((reqd_work_group_size)), [[cl::reqd_work_group_size]], and
+  // [[intel::max_work_group_size]] all require exactly three arguments.
   if ((AL.getKind() == ParsedAttr::AT_ReqdWorkGroupSize &&
        AL.getAttributeSpellingListIndex() ==
            ReqdWorkGroupSizeAttr::CXX11_cl_reqd_work_group_size) ||
-      (AL.getKind() == ParsedAttr::AT_SYCLIntelMaxWorkGroupSize)) {
+      AL.getKind() == ParsedAttr::AT_SYCLIntelMaxWorkGroupSize ||
+      AL.getSyntax() == ParsedAttr::AS_GNU) {
```

Comment:
Can't we just simplify this to check attributes have 3 arguments, since we set the default values for 'optional' arguments prior to this?

Reply:
Or are you keeping it this way because we'll be removing the 'default values' since we have now discovered this is a bug?

Reply:
We can't simplify this check as it's checking the number of args passed to the parsed form of the attribute (so we've not applied any default values to it yet -- that happens when making the semantic attribute).